src/character.cpp

```diff
@@ -1936,6 +1936,7 @@ bool Character::pour_into( item &container, item &liquid )
         auto qty = std::min( liquid.charges, container.ammo_capacity() - container.ammo_remaining() );
         liquid.charges -= qty;
         container.ammo_set( liquid.typeId(), container.ammo_remaining() + qty );
+        container.on_contents_changed();
```

Comment:
> Everything outside of item should use visitable which already includes intrinsic support for nested containers. 

Should, but doesn't do so yet.

Notable cases are:
* Liquid manipulation in `game` and `Character`
* Consumption in `player`
* Advanced inventory (not updated for about a year)

> Within item it's not so easy to enforce and I'm not sure what design would be best

A relatively simple way would be to force non-`const` access to be as explicit as possible. `get_contents` vs. `access_contents` with a big red warning saying "Do not use when not needed".
For more total enforcement, we could even have an opaque nested structure that contains the `contents` and exposes only relevant functions such as `add_content`, `remove_content`, `get_contents (const)`, `access_contents` and invokes a callback on every single non-`const` attempt.